### PR TITLE
LibGfx/CCITT: Add support for Group3 2D

### PR DESCRIPTION
**Pull request by LucasChollet**
Yay, two more versions of the "TIFF :^)" image!

CCITT3 2D images are not often seen in the wild for a good, reason CCITT4 use the same encoding technique but embeds far less bloat so almost everyone use that. Hence, you can see this PR as a big step towards CCITT4 support :^)

### CODE REVIEW

File: Userland/Libraries/LibGfx/ImageFormats/CCITTDecoder.cpp [hunk 1]
```diff
@@ -289,31 +319,9 @@ ErrorOr<void> decode_single_ccitt3_1d_line(BigEndianInputBitStream& input_bit_st
...
-                potential_code = 0;
-            } else if (auto const maybe_terminal = get_terminal_code(current_color, potential_code, size); maybe_terminal.has_value()) {
-                run_length += maybe_terminal->run_length;
-                break;
-            }
-        }
-
-        if (size == 14)
-            return Error::from_string_literal("TIFFImageDecoderPlugin: Invalid CCITT code");
+        current_color = current_color == ccitt_white ? ccitt_black : ccitt_white;
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/CCITTDecoder.cpp:334]: this looks like you're reverting the `invert()` change from the previous commit

File: Tests/LibGfx/TestImageDecoder.cpp [hunk 2]
```diff
@@ -630,6 +630,30 @@ TEST_CASE(test_tiff_ccitt3_fill)
     EXPECT_EQ(frame.image->get_pixel(5, 3), Gfx::Color::NamedColor::White);
 }
 
+TEST_CASE(test_tiff_ccitt3_2d)
+{
+    auto file = TRY_OR_FAIL(Core::MappedFile::map(TEST_INPUT("tiff/ccitt3_2d.tiff"sv)));
```
Comment (nico) [inline Tests/LibGfx/TestImageDecoder.cpp:635]: can you say how you created the test images?

File: Userland/Libraries/LibGfx/ImageFormats/CCITTDecoder.cpp [hunk 3]
```diff
@@ -302,8 +311,11 @@ ErrorOr<u32> read_run_length(BigEndianInputBitStream& input_bit_stream, Color cu
     return run_length;
 }
 
-ErrorOr<void> decode_single_ccitt3_1d_line(BigEndianInputBitStream& input_bit_stream, BigEndianOutputBitStream& decoded_bits, u32 image_width)
+ErrorOr<ReferenceLine> decode_single_ccitt3_1d_line(BigEndianInputBitStream& input_bit_stream, BigEndianOutputBitStream& decoded_bits, u32 image_width)
 {
+    // This is only useful for the 2D decoder
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/CCITTDecoder.cpp:316]: .

File: Userland/Libraries/LibGfx/ImageFormats/CCITTDecoder.cpp [hunk 4]
```diff
@@ -349,6 +361,152 @@ static ErrorOr<void> read_eol(BigEndianInputBitStream& bit_stream, Group3Options
...
+        if (next_is_1D) {
+            reference_line = TRY(decode_single_ccitt3_1d_line(input_bit_stream, decoded_bits, image_width));
+        } else {
+            ReferenceLine current_line {};
+            Color current_color { ccitt_white };
+            u32 column {};
+
+            auto const next_change_on_reference_line = [&](Search search = Search::B1) -> ErrorOr<Change> {
+                // 4.2.1.3.1 Definition of changing picture elements
+                Optional<Change> next_change {}; // This is referred as b1 in the spec
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/CCITTDecoder.cpp:432]: "This is called b1 in the spec" or "This is referred to as b1 in the spec" I think

**Review by nico — APPROVED** [review]
One thing that looks like a merge mistake, otherwise looks great.

File: Tests/LibGfx/TestImageDecoder.cpp (hunk 2 above)
Comment (LucasChollet) [reply Tests/LibGfx/TestImageDecoder.cpp:635]: For once I thought to write it, but I ended fixing up the commit where I put these tests, loosing the explanation in the process :sweat_smile:

**Review by nico — APPROVED** [review]

**Pull request merged**